plugins/ImageActions/main.lua

```diff
@@ -39,11 +42,97 @@ function cb(mode)
       if image:bands()<4 then
         image = image:bandjoin(255) -- add alpha channel
       end
-      white = {240, 240, 240, 0} -- pixels are considered white, if its r,g,b values are >=240 and alpha is arbitrary
-      transparent = {0, 0, 0, 0}
+      local white = {240, 240, 240, 0} -- pixels are considered white, if its r,g,b values are >=240 and alpha is arbitrary
+      local transparent = {0, 0, 0, 0}
       image = image:more(white):bandand():ifthenelse(transparent, image)
+    elseif mode == 8 then
+      local width = image:width()
+      local height = image:height()
+      image = image:crop(image:find_trim())
+      local newWidth = image:width()
+      local newHeight = image:height()
+      maxWidth, maxHeight = maxWidth * newWidth // width, maxHeight * newHeight // height
+    end
+
+    if mode ~= 9 and mode ~=10 then
+      table.insert(imdata, {data = image:write_to_buffer(".png"), x=x, y=y, maxWidth=maxWidth, maxHeight=maxHeight})
+    else
+      local vertical = mode == 9
+      -- find largest sequence of white or transparent pixel rows/columns
+      local filter = image:colourspace("b-w"):extract_band(0):more(240)
+      local width = filter:width()
+      local height = filter:height()
+      local last = vertical and height or width
+      local isWhite = {}
+
+      for i = 0, last - 1 do
+        if vertical then
+          isWhite[i+1] = filter:extract_area(0, i, width, 1):avg() > 254
+        else
+          isWhite[i+1] = filter:extract_area(i, 0, 1, height):avg() > 254
+        end
+      end
+
+      if #isWhite == 0 then
+        word = vertical and "rows" or "columns"
+        print("There are no completely white " .. word .. " in this image. Skipping.") 
+        table.insert(imdata, {data = im["data"], x=x, y=y, maxHeight=maxHeight, maxWidth=maxWidth})
+        goto continue 
+      end
+
+      local whiteEnds = {}
+      local whiteLengths = {}
+      local firstAfter = false
+      local currentLength = 0
+      local n = 0
+      for i = 1, last do
+        if isWhite[i] then
+          if currentLength > 0 then
+            currentLength = currentLength + 1
+          else 
+            currentLength = 1
+            firstAfter = true
+          end
+        else 
+          if firstAfter then
+            whiteEnds[n] = i - 1
+            whiteLengths[n] = currentLength
+            n = n + 1
```

Comment:
There is a significant performance difference (`table.insert` is slower), see https://stackoverflow.com/questions/38345582/why-is-tt-1-e-faster-than-table-insertt-e. I experienced that as well when I optimized the performance for the Notability plugin.
Moreover I start with 0, because the first White block doesn't really count and is discarded later.

Reply:
Oh that's interesting. The we should definitely keep it that way for larger images splitting is already slow and xournalpp freezes.

Still I feel like we should avoid the 0 indexing. I'm also not sure if 0 is in the array part or in the hash part of the table (see the link below).

---
If you're more into optimization maybe https://www.lua.org/gems/sample.pdf is interesting (not sure if everything in it still applies) if you didn't read it yet (it doesn't talk about `table.insert` though).

Reply:
I have managed to increase the speed quite a lot. The bottleneck was the creation of the `isWhite` table, which basically used random access on the elements of the filter. Now it's done via `vips_project()` and by casting the projection into an array.

Reply:
Oh that's nice especially as with this vips function we don't need the whole `isWhite` processing step anymore. The performance difference is definitely notable, nice work :+1: